TextSelection/Views/MultitouchTextView.swift

```diff
@@ -0,0 +1,225 @@
+        
+        if (index > self.highlightSelection!.first) {
+            self.highlightSelection!.first = self.highlightSelection!.start
+            self.highlightSelection!.last = self.textStorage.string.nonWhitespaceCharacterIndexAfter(index)
+        } else {
+            self.highlightSelection!.first = self.highlightSelection!.end
+            self.highlightSelection!.last = self.textStorage.string.nonWhitespaceCharacterIndexBefore(index)
+        }
+        
+        self.updateViews(false)
+        self.selectionDidChange()
+    }
+    
+    func endHighlighting(point: CGPoint) {
+        guard self.highlightSelection != nil else {
+            return
+        }
+        
+        self.preSelection = self.highlightSelection!.add(self.preSelection)
+        self.preSelection?.attributes = [NSForegroundColorAttributeName: preSelectionTextColor, NSBackgroundColorAttributeName: preSelectionColor]
+        self.highlightSelection = nil
+        self.updateViews(false)
+        self.selectionDidChange()
+        
+        self.clearPreSelectionWithDelay()
+    }
+    
+    func appendPreSelectionToSelection() {
+        guard let preSelection = self.preSelection else {
+            return
+        }
+        self.selection = preSelection.add(self.selection)
+        self.selection?.attributes = [NSForegroundColorAttributeName: selectionTextColor, NSBackgroundColorAttributeName: selectionColor]
+        self.preSelection = nil
+        self.updateViews(false)
+        self.selectionDidChange()
+    }
+    
+    func clearSelection() {
+        self.preSelectionTimer?.invalidate()
+        self.highlightSelection = nil
+        self.preSelection = nil
+        self.selection = nil
+        self.updateViews(false)
+        self.selectionDidChange()
+    }
+    
+    func updateViews(animated: Bool) {
+        if (animated) {
+            UIView.transitionWithView(self.textInputView, duration: 0.3, options: [.TransitionCrossDissolve], animations: {
+                self.updateTextAttributes()
+                }) { (done) in
+                    // done
+            }
+        } else {
+            self.updateTextAttributes()
+        }
+    }
+    
+    func updateTextAttributes() {
+        self.textStorage.beginEditing()
+        
+        self.textStorage.setAttributes(self.baseAttributes, range:self.textStorage.fullRange)
+        
+        if let highlightSelection = self.highlightSelection {
+            self.textStorage.addAttributes(highlightSelection.attributes, range:highlightSelection.range)
+        }
+        
+        if let preSelection = self.preSelection {
+            self.textStorage.addAttributes(preSelection.attributes, range:preSelection.range)
+        }
+        
+        if let selection = self.selection {
+            self.textStorage.addAttributes(selection.attributes, range:selection.range)
+        }
+        
+        self.textStorage.endEditing()
```

Comment:
Very subtle, doesn't always cause bugs if you don't but you are supposedly REQUIRED to call begin/end editing. See https://developer.apple.com/library/ios/documentation/StringsTextFonts/Conceptual/TextAndWebiPhoneOS/CustomTextProcessing/CustomTextProcessing.html